### DISCUSSION TITLE
Sections' names in the index using review-index-sample

### DISCUSSION THREAD

**Discussion by Nyoho** [kmuto/review#1869] [Q&A] [answered]
kmutoさんのEPUBで索引を作るサンプル https://github.com/kmuto/review-index-sample がすばらしく、自分なりにいじってみています。

現在、索引でリンクするときの文字列 (通常の紙の索引であればページ番号になるところ) に、□が使われています。

ジャンプする前に、大まかでもいいので、どの辺りに書いてあるのか分かりたいと思い、章番号になるように改造してみました。

コード: https://github.com/Nyoho/review-index-sample/commit/d049640563ef6e0dc0ae069516899f84cae650f9

スクリーンショット: 

<img width="1136" alt="スクリーンショット 2022-11-23 11 58 57" src="https://user-images.githubusercontent.com/118150/203461395-3a916782-e5ad-474e-b31c-af7d5284a4e9.png">

<img width="1136" alt="スクリーンショット 2022-11-23 11 59 02" src="https://user-images.githubusercontent.com/118150/203461401-aeddcfbb-6169-4c9b-a7c5-3caa60a45137.png">

## 現在の方法
- 章番号 format_number を使う
- predef, postdef の場合は format_number がないので、章タイトルを使う

## やりたいこと
これが、 *章単位* ではなく、idx/hdixがある (存在すれば) セクション番号、サブセクション番号 (例えば 2.1 など)が取ってこられるなら、もう少しどの辺りに書いてあるのかが細かくなってうれしいと考えます。

## 質問
(存在すれば) セクション番号、サブセクション番号を索引のリンク文字列にする方法はどういうアプローチを取れば実装できるでしょうか。

**Comment by kmuto** [comment, marked as answer]
なるほど。
Re:VIEWはDOMをまじめに作っていなくてむしろSAX的に頭から解析していくので、ここ(索引)の構造位置はどこ？というのが取りづらいんですよね。

つまり、SAX的に考えるのがよいかと思います。
HTMLBuilderにおいて番号生成および見出しを出力しているメソッド `headline` をオーバーライド(とはいえ `super`と再度 `headline_prefix`  を使うくらい)して、そのときの番号、番号がnilなら見出しを変数に記憶しておく。節までにしておきたいといったことがあればlevelの深さによって覚えるかどうか分岐もしたほうがよいでしょう。
https://github.com/kmuto/review/blob/e7c74313e364e5b994f5ce313ceadd27632d5e84/lib/review/htmlbuilder.rb#L165-L186

hidxに至ったときには記憶した変数から拾えるはずです。

**Comment by Nyoho** [comment]
SAX的に `headline` overriding で覚えておく方法でできました!
code: https://github.com/Nyoho/review-index-sample/commit/978fa7f4d77e0c54f32d52e7b61da426e089a8ff

再度 `headline_prefix` を使うと番号がインクリメントされてしまいましたので、中身の必要なところを取り出して

```ruby
prefix = @sec_counter.prefix(level, @book.config['secnolevel'])
```

としました。
ありがとうございました!
<img width="1136" alt="image" src="https://user-images.githubusercontent.com/118150/204382515-ca280497-2dfe-41f9-a8e9-f588b1bd1714.png">
<img width="1136" alt="image" src="https://user-images.githubusercontent.com/118150/204382549-46870bb9-89e6-445d-b70d-f98644080bd2.png">

**Reply by Nyoho** [reply]
不十分でした。
複数の索引項目がある場合に最後のchap_nameを全部に使ってしまうバグを修正しました。 https://github.com/Nyoho/review-index-sample/commit/1392db9a13aa12b25ceba9cd5337dad7d5803a26
<img width="1136" alt="image" src="https://user-images.githubusercontent.com/118150/204522414-109b55cc-84df-48d3-bc6a-a330b5ebb683.png">
